core/src/apps/webauthn/__init__.py

```diff
@@ -314,141 +423,541 @@ def to_msg(self) -> Msg:
+    try:
+        trezor.pin.keepalive_callback = keepalive_callback
+        if config.has_pin():
+            pin = await request_pin("Enter your PIN", config.get_pin_rem())
+            while config.unlock(trezor.pin.pin_to_int(pin)) is not True:
+                pin = await request_pin("Wrong PIN, enter again", config.get_pin_rem())
+            ret = True
+        else:
+            ret = config.unlock(trezor.pin.pin_to_int(""))
+    except PinCancelled:
+        ret = False
+    finally:
+        trezor.pin.keepalive_callback = None
 
-    def compare(self, action: int, checksum: bytes) -> bool:
-        if self.action != action or self.checksum != checksum:
-            return False
-        if utime.ticks_ms() >= self.deadline:
-            if self.workflow is not None:
-                loop.close(self.workflow)
-            return False
-        return True
+    return ret
 
-    def setup(self, action: int, checksum: bytes, app_id: bytes) -> bool:
-        if workflow.workflows:
-            return False
 
-        self.action = action
-        self.checksum = checksum
-        self.app_id = app_id
+class ConfirmInfo:
+    def __init__(self) -> None:
+        self.app_icon = None  # type: Optional[bytes]
 
-        self.confirmed = None
-        self.workflow = self.confirm_workflow()
-        loop.schedule(self.workflow)
-        return True
+    def get_header(self) -> Optional[str]:
+        return None
+
+    def app_name(self) -> Optional[str]:
+        return None
+
+    def account_name(self) -> Optional[str]:
+        return None
 
-    def keepalive(self):
-        self.deadline = utime.ticks_ms() + _CONFIRM_TIMEOUT_MS
+    def load_icon(self, rp_id_hash: bytes) -> None:
+        from trezor import res
+        from apps.webauthn import knownapps
 
-    async def confirm_workflow(self) -> None:
         try:
-            workflow.onstart(self.workflow)
-            await self.confirm_layout()
-        finally:
-            workflow.onclose(self.workflow)
-            self.workflow = None
+            namepart = knownapps.knownapps[rp_id_hash].lower().replace(" ", "_")
+            icon = res.load("apps/webauthn/res/icon_%s.toif" % namepart)
+        except Exception as e:
+            icon = res.load("apps/webauthn/res/icon_webauthn.toif")
+            if __debug__:
+                log.exception(__name__, e)
+        self.app_icon = icon
+
+
+class ConfirmContent(ui.Control):
+    def __init__(self, info: ConfirmInfo) -> None:
+        self.info = info
+        self.repaint = True
+
+    def on_render(self) -> None:
+        if self.repaint:
+            header = self.info.get_header()
+
+            if header is None or self.info.app_icon is None:
+                return
+            ui.header(header, ui.ICON_DEFAULT, ui.GREEN, ui.BG, ui.GREEN)
+            ui.display.image((ui.WIDTH - 64) // 2, 48, self.info.app_icon)
+
+            app_name = self.info.app_name()
+            account_name = self.info.account_name()
+
```

Comment:
What says CTAP2 and Webauthn about displayed information:

|          | authenticatorMakeCredential                                                                                           | authenticatorGetAssertion                                                                                         |   |   |
|----------|-----------------------------------------------------------------------------------------------------------------------|-------------------------------------------------------------------------------------------------------------------|---|---|
| **CTAP2**    | If the authenticator has a display, show the items contained within the user and rp parameter structures to the user. | Display all these credentials to the user, using their friendly name along with other stored account information. |   |   |
| **Webauthn** | The prompt SHOULD display `rpEntity.id`, `rpEntity.name`, `userEntity.name` and `userEntity.displayName`, if possible.        | Prompt the user to select a public key credential source `selectedCredential` from `credentialOptions`.               |   |   |
|          |                                                                                                                       |                                                                                                                   |   |   |

Reply:
Already discussed.